File: app/src/main/java/info/mx/tracks/util/CommentHelper.java
package info.mx.tracks.util;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.AlertDialog;
import android.content.Context;
import android.provider.Settings;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.EditText;
import android.widget.RatingBar;
import android.widget.Toast;

import com.robotoworks.mechanoid.db.SQuery;
import com.robotoworks.mechanoid.db.SQuery.Op;

import java.util.Locale;

import info.mx.tracks.BuildConfig;
import info.mx.tracks.MxCoreApplication;
import info.mx.tracks.R;
import info.mx.tracks.prefs.MxPreferences;
import info.mx.tracks.sqlite.MxInfoDBContract.Ratings;
import info.mx.tracks.sqlite.RatingsRecord;
import info.mx.tracks.sqlite.TracksRecord;

public class CommentHelper {

    private static RatingsRecord ratingsRecord;
    private static EditText edtComment;
    private static EditText userName;
    private static RatingBar ratingBar;

    public static void doAddRating(Activity activity, final TracksRecord trackRec) {
        // on add eventId is not used
        doEditRating(activity, 0, trackRec);
    }

    private static void doEditRating(final Activity activity, final int ratingId, TracksRecord trackRec) {
        ratingsRecord = RatingsRecord.get(ratingId);
        AlertDialog.Builder builder = new AlertDialog.Builder(activity);
        builder.setTitle(activity.getString(R.string.newRating));

        builder.setNegativeButton(android.R.string.cancel, (dialog, which) -> {
        });

        builder.setPositiveButton(R.string.done, (dialog, which) -> {

            final boolean alreadyExist = SQuery.newQuery()
                    .expr(Ratings.TRACK_REST_ID, Op.EQ, trackRec.getRestId())
                    .expr(Ratings.ANDROIDID, Op.EQ, getAndroidID(activity))
                    .expr(Ratings.NOTE, Op.EQ, edtComment.getText().toString()).exists(Ratings.CONTENT_URI);

            if (!alreadyExist) {
                if (ratingsRecord == null) {
                    ratingsRecord = new RatingsRecord();
                }

                ratingsRecord.setTrackRestId(trackRec.getRestId());
                ratingsRecord.setApproved(0);
                ratingsRecord.setRating((long) ratingBar.getRating());
                ratingsRecord.setUsername(userName.getText().toString());
                ratingsRecord.setNote(edtComment.getText().toString());
                ratingsRecord.setCountry(Locale.getDefault().getCountry());
                ratingsRecord.setChanged(System.currentTimeMillis());
                ratingsRecord.setAndroidid(getAndroidID(activity));
                ratingsRecord.save();
                MxCoreApplication.Companion.doSync(true, true, BuildConfig.FLAVOR);
                if (!ratingsRecord.getUsername().equals(activity.getString(R.string.noname))) {
                    final MxPreferences prefs1 = MxPreferences.getInstance();
                    prefs1.edit().putUsername(ratingsRecord.getUsername().trim()).commit();
                }
            } else {
                Toast.makeText(activity, activity.getString(R.string.comment_alredy_exists), Toast.LENGTH_SHORT).show();
            }
        });
        AlertDialog dlg = builder.create();
        LayoutInflater inflater = dlg.getLayoutInflater();
        @SuppressLint("InflateParams")
        View view = inflater.inflate(R.layout.layout_comment_add, null);
        userName = view.findViewById(R.id.com_username);
        edtComment = view.findViewById(R.id.com_note);
        ratingBar = view.findViewById(R.id.com_rating);
        if (ratingsRecord != null) {
            edtComment.setText(ratingsRecord.getNote());
        }

        final MxPreferences prefs = MxPreferences.getInstance();
        userName.setHint(R.string.default_username);
        userName.setText(prefs.getUsername());
        dlg.setView(view);
        dlg.show();
    }

    private static String getAndroidID(Activity activity) {
        return Settings.Secure.getString(activity.getContentResolver(), Settings.Secure.ANDROID_ID);
    }

    public static void doDeleteNote(final Context context, final int ratingId) {
        final RatingsRecord record = RatingsRecord.get(ratingId);
        AlertDialog.Builder builder = new AlertDialog.Builder(context);

        builder.setPositiveButton(android.R.string.ok, (dialog, which) -> {
            if (record != null) {
                record.delete();
            }
        });
        builder.setNegativeButton(android.R.string.cancel, (dialog, which) -> {
        });
        final AlertDialog delDialog = builder.create();
        delDialog.setMessage("Do you really want to delete this Note?"); // TODO
        delDialog.show();
    }

}

File: app/src/main/java/info/mx/tracks/util/CommentHelper.kt
package info.mx.tracks.util

import android.annotation.SuppressLint
import android.app.Activity
import android.app.AlertDialog
import android.content.DialogInterface
import android.provider.Settings
import android.widget.EditText
import android.widget.RatingBar
import android.widget.Toast
import com.robotoworks.mechanoid.db.SQuery
import info.mx.core.MxCoreApplication.Companion.doSync
import info.mx.core_generated.prefs.MxPreferences.Companion.instance
import info.mx.core_generated.sqlite.MxInfoDBContract.Ratings
import info.mx.core_generated.sqlite.RatingsRecord
import info.mx.core_generated.sqlite.TracksRecord
import info.mx.tracks.BuildConfig
import info.mx.tracks.R
import java.util.Locale

object CommentHelper {
    private var ratingsRecord: RatingsRecord? = null

    // on add eventId is not used
    fun doAddRating(activity: Activity, trackRec: TracksRecord) {
        doEditRating(activity, 0, trackRec)
    }

    private fun doEditRating(activity: Activity, ratingId: Int, trackRec: TracksRecord) {
        var edtComment: EditText? = null
        var userName: EditText? = null
        var ratingBar: RatingBar? = null

        ratingsRecord = RatingsRecord.get(ratingId.toLong())
        val builder = AlertDialog.Builder(activity)
        builder.setTitle(activity.getString(R.string.newRating))

        builder.setNegativeButton(android.R.string.cancel) { _: DialogInterface?, _: Int -> }

        builder.setPositiveButton(R.string.done) { _: DialogInterface?, _: Int ->
            val alreadyExist = SQuery.newQuery()
                .expr(Ratings.TRACK_REST_ID, SQuery.Op.EQ, trackRec.restId)
                .expr(Ratings.ANDROIDID, SQuery.Op.EQ, getAndroidID(activity))
                .expr(Ratings.NOTE, SQuery.Op.EQ, edtComment!!.text.toString())
                .exists(Ratings.CONTENT_URI)
            if (!alreadyExist) {
                if (ratingsRecord == null) {
                    ratingsRecord = RatingsRecord()
                }

                ratingsRecord!!.setTrackRestId(trackRec.restId)
                ratingsRecord!!.setApproved(0)
                ratingsRecord!!.setRating(ratingBar!!.rating.toLong())
                ratingsRecord!!.setUsername(userName!!.text.toString())
                ratingsRecord!!.setNote(edtComment!!.text.toString())
                ratingsRecord!!.setCountry(Locale.getDefault().country)
                ratingsRecord!!.setChanged(System.currentTimeMillis())
                ratingsRecord!!.setAndroidid(getAndroidID(activity))
                ratingsRecord!!.save()
                doSync(true, true, BuildConfig.FLAVOR)
                if (ratingsRecord!!.username != activity.getString(R.string.noname)) {
                    val prefs1 = instance
                    prefs1.edit().putUsername(ratingsRecord!!.username.trim { it <= ' ' })
                        .commit()
                }
            } else {
                Toast.makeText(
                    activity,
                    activity.getString(R.string.comment_alredy_exists),
                    Toast.LENGTH_SHORT
                ).show()
            }
        }
        val dlg = builder.create()
        val inflater = dlg.layoutInflater
        @SuppressLint("InflateParams") val view =
            inflater.inflate(R.layout.layout_comment_add, null)
        edtComment = view.findViewById<EditText>(R.id.com_note)
        ratingBar = view.findViewById<RatingBar>(R.id.com_rating)
        if (ratingsRecord != null) {
            edtComment!!.setText(ratingsRecord!!.note)
        }

        val prefs = instance
        userName = view.findViewById<EditText>(R.id.com_username).apply {
            setHint(R.string.default_username)
            setText(prefs.username)
        }
        dlg.setView(view)
        dlg.show()
    }

    @SuppressLint("HardwareIds")
    private fun getAndroidID(activity: Activity): String? {
        return Settings.Secure.getString(activity.contentResolver, Settings.Secure.ANDROID_ID)
    }
}
